mtsmorf/io/move/read.py

```diff
@@ -0,0 +1,561 @@
+                            _pl(this_len),
+                            type(v).__name__,
+                        )
+                    else:
+                        entr = ""
+            if entr != "":
+                non_empty += 1
+                strs.append("%s: %s" % (k, entr))
+        st = "\n ".join(sorted(strs))
+        st += "\n>"
+        st %= non_empty
+        return st
+
+    def to_data_frame(self):
+        df = pd.DataFrame.from_dict(self.__dict__.items(), orient='columns').T
+        df.columns = df.iloc[0]
+
+        # set datatype
+        df['beh_fname'] = df['beh_fname'].astype(str)
+        df['event_fname'] = df['event_fname'].astype(str)
+
+        df.drop(df.index[0], inplace=True)
+        return df
+
+    def _add_xy_metrics(self, metrics_df):
+        if not self.success or self.perturbed:
+            return
+
+        summ_df = metrics_df.groupby("trial_idx")["Speed"].agg([np.nanmean, np.nanstd])
+        summ_df.reset_index(inplace=True)
+
+        avg_speed = summ_df[summ_df['trial_idx'] == self.trial_idx]['nanmean']
+        std_speed = summ_df[summ_df['trial_idx'] == self.trial_idx]['nanstd']
+        self.avg_speed = avg_speed.values[0]
+        self.std_speed = std_speed.values[0]
+        self.trial_duration = metrics_df[metrics_df['trial_idx'] == self.trial_idx]['trial_duration'].values[0]
+
+
+def compute_xy_metrics(raw, event_key_start, event_key_end,
+                       remove_perturbed=True, verbose: bool = True):
+    if (event_key_start is not None and event_key_end is None) \
+            or (event_key_end is not None and event_key_start is None):
+        raise RuntimeError(f'If event key start/end is set, then '
+                           f'the other also must be set.')
+
+    bids_path = raw.filenames[0].copy().update(
+        recording=None
+    )
+    behav_path = bids_path.copy().update(
+        suffix='behav', extension='.tsv'
+    )
+    events_path = bids_path.copy().update(
+        suffix='events', extension='.tsv'
+    )
+    behav_df = pd.read_csv(behav_path, delimiter="\t", index_col=None)
+
+    # apply preprocessing to obtain Epochs
+    # get the events and events id structure
+    start_key = "Reserved (Start Trial)"
+    events, event_id = mne.events_from_annotations(raw)
+    tlock_event_start = event_id[event_key_start]  # Change time locked event
+    tlock_event_stop = event_id[event_key_end]
+    id_start_key = event_id[start_key]
+
+    drop_inds = _get_bad_epochs(behav_df,
+                                remove_perturbed=remove_perturbed,
+                                remove_unsuccessful=True)
+
+    # indices in event array that Trials start
+    start_indices = np.argwhere(events[:, 2] == id_start_key)
+
+    # samples of the start/end to compute xy coordinate
+    # metrics of interest
+    start_samples = []
+    stop_samples = []
+
+    # now let's loop through these
+    for trial_idx, start_idx in enumerate(start_indices):
```

Comment:
There should be a way to `join` the `behav` and `events` if you cast them to `pd.Dataframe` instances, right? This would reduce the need the for-loop all the time.

Reply:
So the problem w/ the events.tsv I realized is that because some trials are "unsuccessful", they do not have all the event markers. I did this naive brute-force way initially to see how it works.

I'll have to look at the inner join. That may potentially be nice to optimize speed. Rn it's not terribly slow